include/alpaka/queue/cuda_hip/QueueUniformCudaHipRt.hpp

```diff
@@ -0,0 +1,298 @@
+                // has already finished.
+                const auto pCallbackSynchronizationData
+                    = reinterpret_cast<CallbackSynchronizationData*>(arg)->shared_from_this();
+
+                // Notify the executing thread.
+                {
+                    std::unique_lock<std::mutex> lock(pCallbackSynchronizationData->m_mutex);
+                    pCallbackSynchronizationData->m_state = CallbackState::notified;
+                }
+                pCallbackSynchronizationData->m_event.notify_one();
+
+                // Wait for the executing thread to finish the task if it has not already finished.
+                std::unique_lock<std::mutex> lock(pCallbackSynchronizationData->m_mutex);
+                if(pCallbackSynchronizationData->m_state != CallbackState::finished)
+                {
+                    pCallbackSynchronizationData->m_event.wait(
+                        lock,
+                        [pCallbackSynchronizationData]()
+                        { return pCallbackSynchronizationData->m_state == CallbackState::finished; });
+                }
+            }
+
+            ALPAKA_FN_HOST static auto enqueue(
+                uniform_cuda_hip::detail::QueueUniformCudaHipRt<TBlocking>& queue,
+                TTask const& task) -> void
+            {
+                auto pCallbackSynchronizationData = std::make_shared<CallbackSynchronizationData>();
+                ALPAKA_UNIFORM_CUDA_HIP_RT_CHECK(ALPAKA_API_PREFIX(StreamAddCallback)(
+                    queue.getNativeHandle(),
+                    uniformCudaHipRtCallback,
+                    pCallbackSynchronizationData.get(),
+                    0u));
+
+                // We start a new std::thread which stores the task to be executed.
+                // This circumvents the limitation that it is not possible to call CUDA/HIP methods within the CUDA/HIP
+                // callback thread. The CUDA/HIP thread signals the std::thread when it is ready to execute the task.
+                // The CUDA/HIP thread is waiting for the std::thread to signal that it is finished executing the task
+                // before it executes the next task in the queue (CUDA/HIP stream).
+                std::thread t(
+                    [pCallbackSynchronizationData, task]()
+                    {
+                        // If the callback has not yet been called, we wait for it.
+                        {
+                            std::unique_lock<std::mutex> lock(pCallbackSynchronizationData->m_mutex);
+                            if(pCallbackSynchronizationData->m_state != CallbackState::notified)
+                            {
+                                pCallbackSynchronizationData->m_event.wait(
+                                    lock,
+                                    [pCallbackSynchronizationData]()
+                                    { return pCallbackSynchronizationData->m_state == CallbackState::notified; });
```

Comment:
```suggestion
                                    [&pCallbackSynchronizationData]
                                    { return pCallbackSynchronizationData->m_state == CallbackState::notified; });
```
please capture the shared pointer by reference here.

Reply:
I'm not going to, for two reasons:
1. I think it's a bad idea to squeeze unrelated changes in an existing PR/patch set
2. I would need to go through the whole function and lambda hierarchy to understand the lifetime of the various objects and convince myself that this would be a safe change, without any risk that the lifetime of the capturing lambda exceeds that of the captured object; given that these lambdas are being used in different threads, I don't feel confident in doing that

Reply:
On the other hand, @bernhardmgruber if you have done the analysis and convinced yourself that it is a safe, I would suggest you to make this change in a follow up PR.

Reply:
Fine, I'll send a follow-up PR.